### PR TITLE
Adds PyTorch model management

### PR DESCRIPTION
**Pull request by joshdevins**
This is a rough (and working) draft of porting PyTorch code into eland. This is the general shape of what I'm thinking of. It splits concerns into two - (1) generic PyTorch model management in Elasticsearch, and (2) interoperability with HuggingFace transformers and model hub.

Help and comments are welcome. There's surely a lot of formatting and general Python-ification that needs to happen still so help is welcome.

### CODE REVIEW

**Comment by joshdevins** [comment]
PS. I'm happy to just move this work into a branch in `eland` so we can collaborate more easily on this more easily. I suspect it will take some work to get it ready.

File: eland/ml/pytorch/pytorch_model.py
Comment (benwtrent) [inline eland/ml/pytorch/pytorch_model.py:60]: I think the client raises an error if the code isn't a success. So this should be ok.

**Review by benwtrent — APPROVED** [review]
This looks great to me. 

But, I am no Pythonista. Would be good to get a review from @sethmlarson

**Comment by sethmlarson** [comment]
@joshdevins I like your idea of developing this on a branch, I think we can create a branch right off of `main` and then change the target branch for this PR without much fuss. Reviewing this PR now.

File: eland/ml/pytorch/external_models.py
Comment (sethmlarson) [inline eland/ml/pytorch/external_models.py:204]: This looks like real data, are these supposed to be parameters?

File: eland/ml/pytorch/external_models.py
Comment (sethmlarson) [inline eland/ml/pytorch/external_models.py:300]: Seems like a function instead of a staticmethod? Also since there are two distinct functionalities should probably be split into two?

```python
# sort_by_key
sorted_values = [v for _, v in sorted(dict.items(), key=lambda kv: kv[0])]

# sort_by_value
sorted_keys = [k for k, _ in sorted(dict.items(), key=lambda kv: kv[1])]
```

File: eland/ml/pytorch/external_models.py
Comment (sethmlarson) [inline eland/ml/pytorch/external_models.py:323]: This is more readable as `getattr(self._tokenizer, "do_lower_case", False)`

File: eland/ml/pytorch/external_models.py
Comment (sethmlarson) [inline eland/ml/pytorch/external_models.py:354]: nit: Some spacing between each of the `if-elif` blocks here would do a lot for readability.

File: eland/ml/pytorch/pytorch_model.py
Comment (sethmlarson) [inline eland/ml/pytorch/pytorch_model.py:50]: Instead of silently trimming/cleaning up we should raise a `ValueError` with these requirements.

File: eland/ml/pytorch/pytorch_model.py
Comment (sethmlarson) [inline eland/ml/pytorch/pytorch_model.py:80]: TQDM isn't always desirable, only when we're running from CLI. I still haven't "solved" this problem for Eland in general either so not sure we should include here.

File: eland/ml/pytorch/pytorch_model.py
Comment (sethmlarson) [inline eland/ml/pytorch/pytorch_model.py:91]: Is there any cleanup we need to do if one part of the upload process passes and the next fails?

File: eland/ml/pytorch/pytorch_model.py
Comment (sethmlarson) [inline eland/ml/pytorch/pytorch_model.py:116]: For `eland.ml.MLModel` we unconditionally ignore 404s, probably can do that here too?

File: eland/ml/pytorch/pytorch_model.py
Comment (sethmlarson) [inline eland/ml/pytorch/pytorch_model.py:110]: Is there any way to distinguish between the "model deployment is already stopped" and "model isn't found" case? If so might be good to use that here and ignore the "already stopped" case unconditionally and always reraise the 404 model not found case?

File: eland/ml/pytorch/pytorch_model.py
Comment (sethmlarson) [inline eland/ml/pytorch/pytorch_model.py:97]: Timeout should probably be configurable.

File: eland/ml/pytorch/pytorch_model.py
Comment (sethmlarson) [inline eland/ml/pytorch/pytorch_model.py:34]: Are we siloing this to be only PyTorch?

**Review by sethmlarson — CHANGES_REQUESTED** [review]
This is very exciting! I left a whole bunch of review comments. a high level review comment is you can run `nox -rs format` (or `black eland/` will do a lot too) to automatically format code to be compliant with the linter. You'll also get type-checking as well.

**Comment by joshdevins** [comment]
Ok, let me go through the comments here and I will get this into a cleaner state. I'll then push the branch from my fork to this repo so we can collaborate on a branch here until we're ready to merge into master.

File: bin/upload_hub_model.py
Comment (benwtrent) [reply bin/upload_hub_model.py:1]: its an example of how to download a model from hugging face and then deploy it. Its a useful script that the user can run. 

I don't think there is a plan to have it installed on the users system.

Comment (joshdevins) [reply bin/upload_hub_model.py:1]: Sorry this was a copy-paste from the old script so it's incorrect.

I discussed this a couple months ago with @sethmlarson, we should be providing scripts for the most common use-cases so people don't need to write a notebook or their own script. The original intention is indeed that this is installed (on the system/in a venv) with `pip install eland[pytorch]`. Happy to restructure as needed. Also open to any other ideas for how this script should run though.

File: eland/ml/pytorch/external_models.py
Comment (joshdevins) [reply eland/ml/pytorch/external_models.py:193]: That's to conform with the base class which returns `None` by default

File: eland/ml/pytorch/external_models.py
Comment (joshdevins) [reply eland/ml/pytorch/external_models.py:204]: Nope 😄  This does need to be improved but yes, you have to have some hard-coded examples that are used.

File: eland/ml/pytorch/external_models.py
Comment (joshdevins) [reply eland/ml/pytorch/external_models.py:300]: I'm gonna do this change last, because I'm not sure it's possible. Let's come back to this.

File: eland/ml/pytorch/pytorch_model.py
Comment (joshdevins) [reply eland/ml/pytorch/pytorch_model.py:50]: Ok, this is an "automagic" conversion that we need to do, but you're right, it belongs somewhere else (in the place where the ID is automatically created from another model). 💯

Comment (joshdevins) [reply eland/ml/pytorch/pytorch_model.py:50]: Validation is happening already at upload time so I'll leave it there for now

**Comment by sethmlarson** [comment]
@joshdevins The branch has been created and updated the PR base.

File: eland/ml/pytorch/pytorch_model.py
Comment (joshdevins) [reply eland/ml/pytorch/pytorch_model.py:80]: I know, I just don't know how to provide good feedback here to scripts since this upload can take a while. I've seen `autonotebook` used as a good way to control output.

```
from tqdm.autonotebook import tqdm
tqdm.pandas()
```

Comment (joshdevins) [reply eland/ml/pytorch/pytorch_model.py:80]: Or `tqdm.auto` 🤔

File: eland/ml/pytorch/pytorch_model.py
Comment (joshdevins) [reply eland/ml/pytorch/pytorch_model.py:91]: @dimitris-athanasiou ^^ ?

File: eland/ml/pytorch/pytorch_model.py
Comment (joshdevins) [reply eland/ml/pytorch/pytorch_model.py:110]: @benwtrent ^^ ?

File: eland/ml/pytorch/pytorch_model.py
Comment (joshdevins) [reply eland/ml/pytorch/pytorch_model.py:34]: Yes, for now it's all we support.

File: eland/ml/pytorch/external_models.py
Comment (joshdevins) [reply eland/ml/pytorch/external_models.py]: I've made them normal classes but it might be work it's own module. I'm not sure I love as it is now.

File: eland/ml/pytorch/external_models.py
Comment (joshdevins) [reply eland/ml/pytorch/external_models.py]: This is called from outside the class as well so I'm gonna leave it for now. Happy to review once this is on a branch, if there is a better way.

File: eland/ml/pytorch/pytorch_model.py
Comment (benwtrent) [reply eland/ml/pytorch/pytorch_model.py:120]: Looking at the code, `_stop` model deployment doesn't throw when the model is missing or when the deployment is already stopped.

**Pull request merged**
